python/src/aiconfig/editor/server/server.py

```diff
@@ -194,15 +203,42 @@ async def run() -> FlaskResponse:
         params = request_json.get("params", aiconfig.get_parameters(prompt_name))  # type: ignore
         stream = request_json.get("stream", False)
         options = InferenceOptions(stream=stream)
-        run_output = await aiconfig.run(prompt_name, params, options)  # type: ignore
+
+        # Deepcopy the aiconfig prior to run so we can restore it in the case the run operation is cancelled or encounters some error
+        aiconfig_deep_copy = copy.deepcopy(aiconfig)
+
+        # Create a task to run the prompt, and associate the task with the cancellation token
+        # so that we can cancel the task if /cancel is invoked
+        task = asyncio.create_task(aiconfig.run(prompt_name, params, options))  # type: ignore
+        state.tasks[cancellation_token_id] = task  # type: ignore
+
+        # Wait for the task to complete, or for the task to be cancelled
+        run_output = await task  # type: ignore
+
+        # Remove the task from the cancellation token map if the task is complete
+        state.tasks.pop(cancellation_token_id, None)
+
         LOGGER.debug(f"run_output: {run_output}")
         return HttpResponseWithAIConfig(
             #
             message="Ran prompt",
             code=200,
             aiconfig=aiconfig,
         ).to_flask_format()
+    except asyncio.CancelledError:
+        # Restore the aiconfig to its state prior to the run
+        state.aiconfig = aiconfig_deep_copy if aiconfig_deep_copy is not None else state.aiconfig
+
+        # Return a 499 Client Closed Request status code to indicate that the client cancelled the request
+        return HttpResponseWithAIConfig(
+            message="The task was cancelled.",
+            payload={"cancellation_token_id": cancellation_token_id},
+            code=499,
+            aiconfig=state.aiconfig,
+        ).to_flask_format()
     except Exception as e:
+        # TODO: saqadri - should we restore the aiconfig to aiconfig_deep_copy in the case of an exception?
```

Comment:
@saqadri , FYI https://en.wikipedia.org/wiki/Memento_pattern#:~:text=The%20memento%20pattern%20is%20implemented,originator%20for%20a%20memento%20object.

Reply:
Ah I like this pattern. We could add this to Config.py perhaps? (not as part of this change)